### ISSUE TITLE
Nondeterminism Issue

### ISSUE THREAD

**Issue by ben-heil** [greenelab/saged#18]
The CV splits used in the `all_labels` analysis (and probably others) aren't the same between runs. I feel like this could be another set being cast to a list issue, but I'm unsure.

Code to check:
- get_cv_splits
- dataset constructors

**Comment by ben-heil** [comment]
The code below runs deterministically across iterations and across runs

``` python
for i in range(5):                                                                               
        for subset_percent in range(1, 11):                                                          
            subset_percent *= .1                                                                     
                                                                                                     
            # Get CV splits works fine                                                               
            labeled_splits = labeled_data.get_cv_splits(num_splits=args.num_splits, seed=args.seed) 
                                                                                                     
            # Train the model on each fold                                                           
            for i in range(len(labeled_splits)):                                                     
                train_list = labeled_splits[:i] + labeled_splits[i+1:]                               
                                                                                                     
                # Extract the train and test datasets                                                
                LabeledDatasetClass = type(labeled_data)                                             
                train_data = LabeledDatasetClass.from_list(train_list)                               
                train_data = train_data.subset_samples_for_label(subset_percent,                     
                                                                 args.negative_class,                
                                                                 args.seed)                          
                print('Samples: {}'.format(len(train_data.get_samples())))                           
                print('Studies: {}'.format(len(train_data.get_studies())))                           
                val_data = labeled_splits[i]                                                         
            print('')                     
```

**Comment by ben-heil** [comment]
The code below doesn't fail, so the cross-validation splits are actually deterministic. The study names in the output files are in a different order, which is why I thought there was an issue, but the studies themselves are the same across runs. The nondeterminism in the reporting of studies comes because `supervised_train_studies.append(','.join(train_data.get_studies()))` silently casts the study set to a list which is inherently nondeterministic. 

``` python
in_files = glob.glob('results/all_labels*.0.tsv')                                                   
                                                                                                    
sample_sets = [set() for i in range(5)]                                                             
                                                                                                    
for file in in_files:                                                                               
    file_df = pd.read_csv(file, sep='\t')                                                           
                                                                                                    
    for i in range(len(file_df)):                                                                   
        study_str = file_df.iloc[i]['train studies']                                                
        studies = study_str.split(',')                                                              
                                                                                                    
        if len(sample_sets[i]) == 0:                                                                
            sample_sets[i] = set(studies)                                                           
        else:                                                                                       
            assert sample_sets[i] == set(studies) 
```

**Comment by ben-heil** [comment]
The data the models are trained on is the same, but the models themselves yield different sometimes, maybe due to different initializations?

**Comment by ben-heil** [comment]
Solved the mystery! (I hope). Neural networks are initialized to random weight settings, which is something I knew. Unfortunately, I assumed that past me had taken that into account and set all the seeds before initialization happened. It turns out that the seeds were being set *after* model initialization, leading to a different initialization each time.

I fixed the problem in commit `81d0625`, so this issue will be closed in the next PR.
